core/src/cluster_info.rs

```diff
@@ -1747,122 +1748,117 @@ impl ClusterInfo {
         requests: Vec<PullData>,
         stakes: &HashMap<Pubkey, u64>,
     ) -> Option<Packets> {
-        // split the requests into addrs and filters
-        let mut caller_and_filters = vec![];
-        let mut addrs = vec![];
         let mut time = Measure::start("handle_pull_requests");
         self.update_data_budget(stakes);
-        for pull_data in requests {
-            caller_and_filters.push((pull_data.caller, pull_data.filter));
-            addrs.push(pull_data.from_addr);
-        }
-        let now = timestamp();
-        let self_id = self.id();
 
-        let pull_responses = self
-            .time_gossip_read_lock(
-                "generate_pull_responses",
-                &self.stats.generate_pull_responses,
-            )
-            .generate_pull_responses(&caller_and_filters, now);
+        let now = Instant::now();
+        let timestamp = timestamp();
+        let num_requests = requests.len();
+        let requests = self.reorder_pull_requests(now, requests, stakes);
 
-        self.time_gossip_write_lock("process_pull_reqs", &self.stats.process_pull_requests)
-            .process_pull_requests(caller_and_filters, now);
-
-        // Filter bad to addresses
-        let pull_responses: Vec<_> = pull_responses
-            .into_iter()
-            .zip(addrs.into_iter())
-            .filter_map(|(responses, from_addr)| {
-                if !from_addr.ip().is_unspecified()
-                    && from_addr.port() != 0
-                    && !responses.is_empty()
-                {
-                    Some((responses, from_addr))
-                } else {
-                    None
-                }
-            })
-            .collect();
+        let mut total_bytes = 0;
+        let mut sent_requests = 0;
+        let mut packets = Packets::new_with_recycler(recycler.clone(), 64, "handle_pull_requests");
 
-        if pull_responses.is_empty() {
-            return None;
+        for request in &requests {
+            if request.from_addr.ip().is_unspecified() || request.from_addr.port() == 0 {
+                continue; // Skip bad addresses.
+            }
+            let (responses, done) = self.make_reponse_packets(request, timestamp);
+            let bytes: usize = responses.iter().map(|packet| packet.meta.size).sum();
+            for packet in responses {
+                packets.packets.push(packet);
+            }
+            self.bandwidth_usage
+                .write()
+                .unwrap()
+                .add(now, request.caller.pubkey(), bytes as u64);
+            total_bytes += bytes;
+            if !done {
+                inc_new_counter_info!("gossip_pull_request-no_budget", 1);
```

Comment:
These logs can get spammy if we're hitting the budget limit, can we fold these into the `cluster_info.stats` that are reported through `cluster_info.print_reset_stats()`

Reply:
done.